bin_strat/indicators.py

```diff
@@ -0,0 +1,94 @@
+from enum import IntFlag
+import pandas as pd
+import numpy as np
+
+class TradingSignal(IntFlag):
+    BUY = 1
+    SELL = -1
+    HOLD = 0
+
+def macd_signal(price_history):
+    """
+    Calculates the most recent macd signal for a given asset
+    - Calculate the MACD and signal lines
+    - Generate the trading signal based off the lines crossing
+
+    For when we can do multiple orders/tickers:
+    - Screener itteration through multiple tickers to compare the Weighted Success Rate
+    - Checking the Binance API account balance for the current balance of the asset
+    Args:
+        price_history (pd.DataFrame): the historical price data for a given asset
+    Returns:
+        TradingSignal: the signal to sell (TradingSignal.SELL), buy (TradingSignal.BUY), or do nothing (TradingSignal.HOLD)
+    """
+    price_history_macd = _MACD_calc(price_history)
+    macd_signal = _MACD_strat(price_history_macd)
+    latest_signal = macd_signal['buy_sell_signal'].iloc[-1]
+
+    return latest_signal
+
+
+def _MACD_calc(price_history):
+    """
+    Calculates the MACD line and signal line
+    Args: price_history (pd.DataFrame): the historical price data for a given asset
+    Returns: price_history_macd (pd.DataFrame) with  columns ['Close', 'ret_pct_change', 'MACD', 'signal', 'go_long', 'potential_gains']
+    """
+    k = price_history['Close'].ewm(span=10, adjust=False).mean() # Get the 26-day EMA of the closing price
+    d = price_history['Close'].ewm(span=19, adjust=False).mean() # Get the 12-day EMA of the closing price
+    macd = k - d # Subtract the 26-day EMA from the 12-Day EMA to get the MACD
+    macd_s = macd.ewm(span=6, adjust=False).mean() # Get the 9-Day EMA of the MACD for the Trigger line
+    price_history_macd = price_history
+    price_history_macd['MACD'] = macd
+    price_history_macd['signal'] = macd_s
+
+    return price_history_macd
+
+
+def _MACD_strat(price_history_macd):
+    """
+    Function to create the buy_sell_signal: 0 for hold, 1 for buy, -1 for sell. Will use TradingSignal
+    Logic: if MACD > signal then buy, if MACD < signal then sell
+    args: price_history_macd (pd.DataFrame)
+    returns: macd_signal (pd.DataFrame)
+    """
+    flag = 0 
+    buy_sell_signal = []
+    entry_price = []
+    exit_price = []
+    for i in range(0,len(price_history_macd)):
+        if price_history_macd.MACD[i] > price_history_macd.signal[i]: #buy if previous indicator is hold(0)
+            if flag != 1:
```

Comment:
Is this equivalent to `if flag not TradingSignal.BUY:`? That would be more human readable to someone not familiar with -1, 0, 1 as trading signals

Reply:
no, so before there was a separate function for 'go-long' that looped through the same data to make 1s and 0s to track whether the previous signals had already bought. I'm using 'flag' here as a tracker for if the security was already bought where TradingSignal could be a 0 but holding onto a buy already or holding onto a sell.

Reply:
So it's being used to track whether the last "action" so to speak was a buy or sell, if I'm understanding correctly?

Reply:
right, 100. The flag is needed because the previous interval may or may not have included an action.